src/Http/Http.Extensions/src/UriHelper.cs

```diff
@@ -210,18 +209,48 @@ public static string GetDisplayUrl(this HttpRequest request)
         var path = request.Path.Value ?? string.Empty;
         var queryString = request.QueryString.Value ?? string.Empty;
 
-        // PERF: Calculate string length to allocate correct buffer size for StringBuilder.
         var length = scheme.Length + SchemeDelimiter.Length + host.Length
             + pathBase.Length + path.Length + queryString.Length;
 
-        return new StringBuilder(length)
-            .Append(scheme)
-            .Append(SchemeDelimiter)
-            .Append(host)
-            .Append(pathBase)
-            .Append(path)
-            .Append(queryString)
-            .ToString();
+        return string.Create(
```

Comment:
Is this meaningfully better than just calling the span `Concat` overload?

```c#
return string.Concat((ReadOnlySpan<string?>)[
    request.Scheme ?? string.Empty,
    SchemeDelimiter,
    request.Host.Value ?? string.Empty,
    request.PathBase.Value ?? string.Empty,
    request.Path.Value ?? string.Empty,
    request.QueryString.Value ?? string.Empty
]);
```

Reply:
Because `string.Concat` with `ReadOnlySpan<char>` only goes up to 5 parameters and 6 are needed, as you figured out, an array allocation is needed (the API you "used" doesn't exist).

I haven't benchmarked this one, but I don't expect it to be better than `string.Create`.

Reply:
It does exist in .Net 9, might be fairly recent.

Reply:
Still can't see how that could be better than `string.Create`. have you benchmarked it?

Reply:
It's better by way of being 10x shorter and thus more readable/maintainable.

Unless rolling out the custom Concat is meaningfully faster, I don't think it's worth the extra logic.
Judging by the implementation, I would expect the two to behave very similarly.

Reply:
> It's better by way of being 10x shorter and thus more readable/maintainable.

It's around 3 times faster and uses around 4 times less memory and is a pattern already used in the code base.

> Unless rolling out the custom Concat is meaningfully faster, I don't think it's worth the extra logic.
Judging by the implementation, I would expect the two to perform very similarly.

Have you benchmarked it?

Reply:
> It's around 3 times faster and uses around 4 times less memory and is a pattern already used in the code base.

I'm not saying we shouldn't make any changes here. Thank you for looking into this and improving things.
Going through `string.Create` definitely is faster than `StringBuilder`, that isn't being disputed.

I'm saying that using an existing helper `string.Concat(ReadOnlySpan<string>)` should be very similar perf-wise to the significantly more logic needed to go through `string.Create`.

Reply:
E.g. for a sample input from your benchmark

| Method               | Mean      | Allocated |
|--------------------- |----------:|----------:|
| String_Interpolation | 111.77 ns |     216 B |
| Concat               |  88.82 ns |     216 B |
| String_Create        |  75.50 ns |     216 B |

Then the question becomes whether the extra LOC are worth it for 10 ns on `GetDisplayUrl`.
I'll leave that up to the maintainers of this repo.

Reply:
Seems odd that `Concat` is slower, it should basically be doing the same thing as `string.Create` here.

Is it possible that the [`CopyStringContent`](https://source.dot.net/#System.Private.CoreLib/src/libraries/System.Private.CoreLib/src/System/String.Manipulation.cs,428) calls are adding overhead that could be reduced?

Reply:
That helper is most likely getting inlined. My guess would be it's a combination of
- extra branches to guard against null values (both when calculating length and before copying)
- extra branch to check for length overflow (which theoretically the manual `string.Create` should also be doing)
- extra branches for length checks before copying (defensive in case the backing values changed)
- extra branch at the end to check if the length is correct
- One extra Memmove call for `"://"` which would otherwise be turned into a couple `mov`s when used directly in a `CopyTo` like in the `string.Create` impl

Reply:
@BrennanConroy,

> Seems odd that `Concat` is slower, it should basically be doing the same thing as `string.Create` here.
> 
> Is it possible that the [`CopyStringContent`](https://source.dot.net/#System.Private.CoreLib/src/libraries/System.Private.CoreLib/src/System/String.Manipulation.cs,428) calls are adding overhead that could be reduced?

[public static string Concat(/*params*/ ReadOnlySpan<string?> values)](https://source.dot.net/#System.Private.CoreLib/src/libraries/System.Private.CoreLib/src/System/String.Manipulation.cs,375) is not yet in .NET 9.0.0-preview.3.24172.9. As soon as it is in a preview,

The current implementation of `string.Concat` that does not require the allocation of a `string[]` only accepts up to 4 `string`s, and this requires 5.

I'll update my benchmarks to use it.

@MihaZupan,

> That helper is most likely getting inlined. My guess would be it's a combination of
> 
> * extra branches to guard against null values (both when calculating length and before copying)
> * extra branch to check for length overflow (which theoretically the manual `string.Create` should also be doing)
> * extra branches for length checks before copying (defensive in case the backing values changed)
> * extra branch at the end to check if the length is correct
> * branches from having the loops at all instead of being effectively manually unrolled
> * One extra Memmove call for `"://"` which would otherwise be turned into a couple `mov`s when used directly in a `CopyTo` like in the `string.Create` impl

When hot path performance is involved, I tend to not rely on guessing. I rely on personal and community experience and concrete measurements.